[doc] Provisioning Admission Check Controller (#1242)
kubernetes-sigs/kueue: site/content/en/docs/admission-check-controllers/provisioning.md

---

The Provisioning Admission Check Controller is an Admission Check Controller designed to integrate Kueues with [Kubernetes cluster-autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler). Its primary function is to create [ProvisioningRequests](https://github.com/kubernetes/autoscaler/blob/4872bddce2bcc5b4a5f6a3d569111c11b8a2baf4/cluster-autoscaler/provisioningrequest/apis/autoscaling.x-k8s.io/v1beta1/types.go#L41) for the workloads holding q [Quota Reservation](docs/concepts/#quota-reservation) and keeping the [AdmissionCheckState](/docs/concepts/admission_check/#admissioncheckstate) in sync.

The controller is part of kueue. You can enable it by setting the `ProvisioningACC` feature gate. Check the [Installation](/docs/installation/#change-the-feature-gates-configuration) guide for details on feature gate configuration.

## Parameters

This controller uses a `ProvisioningRequestConfig` as parameters, like:

```yaml
apiVersion: kueue.x-k8s.io/v1beta1
kind: ProvisioningRequestConfig
metadata:
  name: prov-test-config
spec:
  provisioningClassName: queued-provisioning.gke.io
  managedResources:
  - nvidia.com/gpu
```

Where:

> - **provisioningClassName** - describes the different modes of provisioning the resources. Check autoscaling.x-k8s.io ProvisioningRequestSpec.ProvisioningClassName for details.

Reviewer: ```suggestion
- **provisioningClassName** - describes the different modes of provisioning the resources. Check `autoscaling.x-k8s.io` `ProvisioningRequestSpec.provisioningClassName` for details.
```

---
title: "Provisioning Admission Check Controller"
date: 2023-10-23
weight: 1
description: >
  An admission check controller providing kueue integration with cluster autoscaler.
---

> The Provisioning Admission Check Controller is an Admission Check Controller designed to provide Kueues integration with [Kubernetes cluster autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler). It's mainly focused on creating [ProvisioningRequests](https://github.com/kubernetes/autoscaler/blob/4872bddce2bcc5b4a5f6a3d569111c11b8a2baf4/cluster-autoscaler/provisioningrequest/apis/autoscaling.x-k8s.io/v1beta1/types.go#L41) for the workloads holding [Quota Reservation](docs/concepts/#quota-reservation) and keeping the [AdmissionCheckState](/docs/concepts/admission_check/#admissioncheckstate) in sync.

Reviewer: ```suggestion
The Provisioning Admission Check Controller is an Admission Check Controller designed to integrate Kueue with the [Kubernetes cluster-autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler). It's primary function is to create [ProvisioningRequests](https://github.com/kubernetes/autoscaler/blob/4872bddce2bcc5b4a5f6a3d569111c11b8a2baf4/cluster-autoscaler/provisioningrequest/apis/autoscaling.x-k8s.io/v1beta1/types.go#L41) for the workloads holding a [Quota Reservation](docs/concepts/#quota-reservation) and keeping the [AdmissionCheckState](/docs/concepts/admission_check/#admissioncheckstate) in sync.
```

> The controller is part of kueue and configured by `ProvisioningACC` feature gate, check the [Installation](/docs/installation/#change-the-feature-gates-configuration) guide on details on feature gate configuration.

Reviewer: ```suggestion
The controller is part of kueue. You can enable it by setting the `ProvisioningACC` feature gate. Check the [Installation](/docs/installation/#change-the-feature-gates-configuration) guide on details on feature gate configuration.
```

## Parameters

This controller uses a `ProvisioningRequestConfig` as parameters, which uses the following schema:

> ```yaml

Reviewer: The OpenAPI schema is not human-friendly.

Let's start with the example and follow by describing the important fields.

Author: done

Reviewer: > What will really help is to also publish some kind of generated API documentation.

I agree. Maybe we can investigate how to do that in the next release. That would help a lot.

Reviewer: https://github.com/kubernetes-sigs/kueue/issues/1249

                    in a ResourceList.
                  type: string
                maxItems: 100
                type: array
                x-kubernetes-list-type: set
              parameters:
                additionalProperties:
                  description: Parameter is limited to 255 characters.
                  maxLength: 255
                  type: string
                description: Parameters contains all other parameters classes may
                  require.
                maxProperties: 100
                type: object
              provisioningClassName:
                description: ProvisioningClassName describes the different modes of
                  provisioning the resources. Check autoscaling.x-k8s.io ProvisioningRequestSpec.ProvisioningClassName
                  for details.
                maxLength: 253
                pattern: ^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$
                type: string
            required:
            - provisioningClassName
            type: object
        type: object
```

## Example

### Setup

{{< include "/examples/provisioning/provisioning-setup.yaml" "yaml" >}}

> ### Job

Reviewer: ```suggestion
### Job using a ProvisioningRequest
```

Author: done
